src/charm.py

```diff
@@ -133,7 +149,25 @@ def _reconcile(self, _: ops.EventBase) -> None:
             self.container.start("exporter")
         else:
             self.container.stop("exporter")
-        self.unit.status = ops.ActiveStatus()
+        deadline = time.time() + 300
+        self.unit.status = ops.WaitingStatus("waiting for penpot services")
+        while time.time() < deadline:
+            if self._check_penpot_backend_ready():
+                self.unit.status = ops.ActiveStatus()
+                return
+            time.sleep(3)
+        raise TimeoutError("Timeout waiting for penpot services")
```

Comment:
Could this timeout be generated by a case similar to [this](https://bugs.launchpad.net/juju/+bug/2063087)? If a case like that could happen, it may be better to set the unit to "blocked" instead of raising?

Reply:
The reason for the wait here is that after the Penpot backend starts, it takes about 1 minute before the backend can respond to requests. This is somewhat annoying as it confuses users (the Penpot webpage will show errors) and makes integration testing cumbersome.

At this point, all data from Juju required to run Penpot is available; we are merely waiting for Penpot to start. If it doesn't start, there must be an issue within Penpot that is beyond our control, and there's nothing we can do about it.

Reply:
Oh yeah, you are absolutely right. In that situation, penpot will enter error loop, thus never receiving the relation-changed event, and never being able to recover. Let me update this to enter the blocked state, thanks!

Reply:
What bothers me is that "BlockedState" means we expect an action from the user. If we know we're waiting for some relation data, and the relation is there, we should put in "Maintenance" (like suggested for oauth).
I'm also not a big fan of such wait within the charm, can't we use a pebble check with level "ready" ? (my guess is that since our ingress ignores the readiness probes and still send traffic we can't). We could also use pebble notices by having a startup script that detects when the workload is effectively ready and send the notice. Meanwhile the charm stays in "MaintenanceState"
Note that the startup script could also detect if the startup failed and send another notice.
I'd love pebble to automatically send notices when an heathcheck fails or when the workload is restart... but doesn't seem to be the case for now (to be discussed with CharmTech ? @jdkandersson)